feat(allocator): `TakeIn` trait

overlookmotel opened the pull request:
Add 2 traits and implement them for AST types:

* `Dummy` trait creates a dummy node for an AST type.
* `TakeIn` replaces a node with a dummy.

Reason for 2 separate traits is that for some types we need to implement `Dummy`, but it'd be inappropriate to implement `TakeIn`. e.g.:

* Fieldless enums which are `Copy`: They should just be copied, rather than replaced.
* `Span`: Ditto.
* `Box<T>`: You should replace the *contents* of the `Box`, not the `Box` itself.

Ideally we wouldn't be implementing `TakeIn` for so many types. Some types are very expensive to generate a dummy for, so we'd ideally guide consumers away from that by not implementing `TakeIn` for those types.

But we want to replace the `TakeIn` trait in Rolldown with our own version. Because Rolldown implements `TakeIn` on many many types, it seemed to make sense to codegen `TakeIn` for pretty much all AST types, so we can migrate their code over to using our trait without too much hassle.

overlookmotel commented:
* **#9969** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9969?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9969?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9969).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen reviewed (commented):
I'll try and replace Rolldown's adhoc `TakeIn` trait with ours. Please let me know when.

overlookmotel commented:
@Boshen I think this is ready. It's slightly different from Rolldown's version in that Rolldown implements `TakeIn` for `Box<T>`, which I think is a bad idea. But it should be possible to replace their `TakeIn` with ours with a few tweaks to their code.

Boshen commented:
Can we also feature gate this, other projects other than rolldown require this.

Boshen commented:
Done in https://github.com/rolldown/rolldown/compare/next-oxc?expand=1

@overlookmotel merge after merge conflict.

overlookmotel commented:
### Merge activity

* **Mar 24, 6:27 AM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Mar 24, 6:27 AM EDT**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

overlookmotel committed:
feat(allocator): `TakeIn` trait (#9969)

Add 2 traits and implement them for AST types:

* `Dummy` trait creates a dummy node for an AST type.
* `TakeIn` replaces a node with a dummy.

Reason for 2 separate traits is that for some types we need to implement `Dummy`, but it'd be inappropriate to implement `TakeIn`. e.g.:

* Fieldless enums which are `Copy`: They should just be copied, rather than replaced.
* `Span`: Ditto.
* `Box<T>`: You should replace the *contents* of the `Box`, not the `Box` itself.

Ideally we wouldn't be implementing `TakeIn` for so many types. Some types are very expensive to generate a dummy for, so we'd ideally guide consumers away from that by not implementing `TakeIn` for those types.

But we want to replace the `TakeIn` trait in Rolldown with our own version. Because Rolldown implements `TakeIn` on many many types, it seemed to make sense to codegen `TakeIn` for pretty much all AST types, so we can migrate their code over to using our trait without too much hassle.
